Add RemoveBlock API to allow `DELETE /{index}/_block/{block}`

HyunsangHan opened the pull request:
## Summary
Introduces a new `RemoveBlock` API that complements the existing `AddBlock` API by allowing users to remove index blocks using `DELETE /{index}/_block/{block}`.

Resolves: https://github.com/elastic/elasticsearch/issues/128966


## Background
Currently, users add blocks via the `AddBlock` API (`PUT /{index}/_block/{block}`) but must remove them through the more generic Settings API (`PUT /{index}/_settings`).

## What’s Changed

### Core

* Introduced new classes: `RemoveIndexBlockRequest`, `RemoveIndexBlockResponse`, `TransportRemoveIndexBlockAction`.
* Added `removeIndexBlock` to `MetadataIndexStateService`.

### REST API

* New endpoint: `DELETE /{index}/_block/{block}`
* Added new API spec: `indices.remove_block.json`
* Response uses `AcknowledgedResponse` (no shard-level confirmation required)

### Features

* Integrated with security and operator privileges.
* Safe to re-run (idempotent behavior).

### Compatibility

* Fully backward compatible
* Existing Settings API remains unchanged
* AddBlock API untouched

### Testing

* Unit tests for service logic
* Integration tests covering full API usage and edge cases
* REST YAML tests for endpoint validation
* Security tests for permission checks

HyunsangHan committed:
Add RemoveBlock core service implementation

- Add RemoveIndexBlockRequest/Response/ClusterStateUpdateRequest classes
- Add removeIndexBlock method to MetadataIndexStateService

Provides core service infrastructure for removing index blocks.

HyunsangHan committed:
Add RemoveBlock transport layer and client APIs

- Add TransportRemoveIndexBlockAction and RemoveIndexBlockRequestBuilder
- Add prepareRemoveBlock() and removeBlock() methods to IndicesAdminClient
- Register RemoveBlock actions in ActionModule
- Add integration tests and security permissions

HyunsangHan committed:
Add REST interface for RemoveBlock API

- Add RestRemoveIndexBlockAction for DELETE /{index}/_block/{block}
- Add indices.remove_block.json API specification
- Update YAML REST tests and documentation
- Remove shards_acknowledged field references from docs

Completes public HTTP interface providing symmetry with addBlock API.

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

nielsbauman committed:
Move cluster task logic to transport action

nielsbauman committed:
Use `ProjectState`

nielsbauman commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/indices.blocks/10_basic.yml:
```diff
@@ -24,7 +24,16 @@
         index: test_index
 
   - do:
-      indices.put_settings:
```
Instead of changing the existing test, can you add a new test that uses the remove block API? We don't have any plans yet to remove the existing behavior (of using the settings API), so we still want to cover that in these tests.

nielsbauman replied:
Also make sure to add a newline at the end of this file when you update it.

HyunSangHan replied:
It sounds better! 👍

nielsbauman commented on server/src/main/java/org/elasticsearch/action/admin/indices/readonly/RemoveIndexBlockResponse.java:
```diff
@@ -0,0 +1,146 @@
+
+    public static final RemoveIndexBlockResponse EMPTY = new RemoveIndexBlockResponse(true, List.of());
+
+    private final List<RemoveBlockResult> indices;
```
Can you rename this field (and all related parameters below) to `results`?

nielsbauman reviewed (commented):
Hi @HyunSangHan, thank you very much for your contribution to Elasticsearch! The PR is already looking good, I just have a few small comments for now - I need to have another look later on. I also pushed some changes myself, as they're hard to explain, but I didn't want to completely hijack your PR, so I'll let you take care of the remaining changes.

nielsbauman commented:
buildkite test this

HyunsangHan committed:
Restore settings test and add new remove block test

HyunsangHan committed:
Edit variable name and remove unnecessary overrides

HyunsangHan committed:
Pass timeout parameters directly in RemoveIndexBlockRequest constructor

Replace implicit default timeouts with explicit parameters in constructor
to improve code clarity and follow established patterns.

nielsbauman committed:
Make request cancelleable

nielsbauman commented on server/src/internalClusterTest/java/org/elasticsearch/operateAllIndices/DestructiveOperationsIT.java:
```diff
@@ -155,4 +155,40 @@ public void testAddIndexBlockDefaultBehaviour() throws Exception {
+        // Add blocks first
+        assertAcked(indicesAdmin().prepareAddBlock(WRITE, "index1", "1index").get());
+
+        // Should succeed, since no wildcards
+        assertAcked(indicesAdmin().prepareRemoveBlock(WRITE, "1index").get());
+        // Special "match none" pattern succeeds, since non-destructive
+        assertAcked(indicesAdmin().prepareRemoveBlock(WRITE, "*", "-*").get());
+
+        expectThrows(IllegalArgumentException.class, indicesAdmin().prepareRemoveBlock(WRITE, "i*"));
+        expectThrows(IllegalArgumentException.class, indicesAdmin().prepareRemoveBlock(WRITE, "_all"));
```
Can you switch the order - i.e. first do the rejected requests? Otherwise those requests don't really make sense (on non existing indices).

nielsbauman reviewed (commented):
I just noticed I didn't actually push my commit last time, so I just did so. I added a few more comments. Thanks for incorporating my previous comments so soon!

HyunSangHan commented on server/src/internalClusterTest/java/org/elasticsearch/blocks/SimpleBlocksIT.java:
```diff
@@ -549,6 +550,143 @@ public static void disableIndexBlock(String index, APIBlock block) {
+    public void testRemoveBlockToMissingIndex() {
+        IndexNotFoundException e = expectThrows(
+            IndexNotFoundException.class,
+            indicesAdmin().prepareRemoveBlock(randomAddableBlock(), "test")
```
Thanks for pointing this out!
The `prepareRemoveBlock` method signature was updated but I missed updating the test calls accordingly. I will fix it now!

HyunsangHan committed:
Fix broken test cases using prepareRemoveBlock

HyunsangHan committed:
Update remove_block API test to use capabilities feature

HyunsangHan committed:
Remove redundant description in index-block.md

HyunsangHan committed:
Add new line to the end of indices.remove_block.json

HyunsangHan committed:
Reorder test assertions in testRemoveIndexBlockIsRejected

HyunsangHan committed:
Fix bug for getting block name

Fixed inconsistent block name retrieval that caused uppercase/lowercase mismatch
- Remove block used enum.name() returning uppercase
- Add block used field access returning lowercase
- Override APIBlock.toString() to consistently return lowercase block name

nielsbauman commented:
buildkite test this

nielsbauman commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java:
```diff
@@ -233,7 +286,19 @@ private static ClusterStateListener blockedIndexCountListener() {
                 .stream()
                 .filter(indexMetadata -> INDEX_BLOCKS_WRITE_SETTING.get(indexMetadata.getSettings()))
                 .count(),
-            oneOf(0L, 3L)
+            oneOf(0L, 1L, 2L, 3L)  // Allow intermediate states during batched processing
```
You changed the value of `blockedIndexCountListener`. Is that intentional? I would assume that value needs to be unchanged.

HyunsangHan committed:
Fix assertion for blockedIndexCountListener

HyunsangHan committed:
Remove unused constant

HyunSangHan commented:
@nielsbauman 
After you trigger the build, most of the CI checks failed, even though I didn’t make any changes to those files.
Could it be due to flaky tests, or is there something in my changes that might have caused it? 🤔

nielsbauman committed:
Merge branch 'main' into api-remove-index-block

nielsbauman commented:
buildkite test this

nielsbauman committed:
Add changelog

nielsbauman committed:
Merge branch 'main' into api-remove-index-block

nielsbauman reviewed (commented):
Hi @HyunSangHan, sorry for the late reply. I was discussing the changes with some colleagues who are more familiar with the inner workings of blocks. I added two more small comments, after that I think we're good to go.

HyunsangHan committed:
Remove VERIFIED_READ_ONLY_SETTING when write blocks are cleared

- Add logic to MetadataIndexStateService.removeIndexBlock() to automatically clear VERIFIED_READ_ONLY_SETTING when no write-blocking constraints remain

HyunSangHan commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -1157,6 +1159,88 @@ public static ClusterBlock createUUIDBasedBlock(ClusterBlock clusterBlock) {
+
+                    projectBuilder.put(updatedMetadata, true);
+                }
+
```
That's good catch. Thank you for the detailed review.

HyunSangHan commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -1157,6 +1159,88 @@ public static ClusterBlock createUUIDBasedBlock(ClusterBlock clusterBlock) {
+                    blocks.removeIndexBlockWithId(projectState.projectId(), index.getName(), block.block.id());
+                } else {
+                    blocks.removeIndexBlock(projectState.projectId(), index.getName(), block.block);
+                }
```
I got it. Let me push an updated commit shortly.

nielsbauman commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexStateService.java:
```diff
@@ -1157,6 +1159,83 @@ public static ClusterBlock createUUIDBasedBlock(ClusterBlock clusterBlock) {
+
+                final Settings indexSettings = indexMetadata.getSettings();
+                final boolean hasBlockSetting = block.setting().get(indexSettings);
+
+                boolean hasAnyBlock = hasBlockSetting;
+
+                final Set<ClusterBlock> clusterBlocks = projectState.blocks().indices(projectState.projectId()).get(index.getName());
+                if (clusterBlocks != null) {
+                    for (ClusterBlock clusterBlock : clusterBlocks) {
+                        if (clusterBlock.id() == block.block.id()) {
+                            hasAnyBlock = true;
+                            break;
+                        }
+                    }
+                }
+
+                if (hasAnyBlock == false) {
```
```suggestion

                final boolean hasBlock = projectState.blocks().hasIndexBlock(projectState.projectId(), index.getName(), block.block);
                if (hasBlockSetting == false && hasBlock == false) {
```

nielsbauman committed:
Refactor `hasBlock`

nielsbauman committed:
Merge branch 'main' into api-remove-index-block

nielsbauman commented:
buildkite test this

nielsbauman reviewed (approved):
LGTM! Thank you very much for your contribution and iterations on this PR, @HyunSangHan!

HyunSangHan commented:
@nielsbauman 
Thank you! 🙌 I really enjoyed working through the iterations and learned a lot from the process.
Appreciate all your feedback and support!

HyunSangHan commented:
@nielsbauman 
Hi, Niels!
I noticed that this feature is already reflected in the official documentation.
I've opened an issue to report it: https://github.com/elastic/elasticsearch/issues/130242

HyunSangHan commented:
We can find RemoveBlock API in the document after recent release:
https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-indices-remove-block
